R/data_modify.R

```diff
@@ -0,0 +1,245 @@
+      })
+      names(dots) <- vapply(dots, function(n) insight::safe_deparse(n[[2]]), character(1))
+    }
+  }
+
+  for (i in seq_along(dots)) {
+    # iterate expressions for new variables
+    symbol <- dots[[i]]
+
+    # expression is given as character string in a variable, but named, e.g.
+    # a <- "2 * Sepal.Width"
+    # data_modify(iris, double_SepWidth = a)
+    # we reconstruct the symbol as if it were provided as literal expression.
+    # However, we need to check that we don't have a character vector,
+    # like: data_modify(iris, new_var = "a")
+    # this one should be recycled instead.
+    if (!is.character(symbol)) {
+      eval_symbol <- .dynEval(symbol, ifnotfound = NULL)
+      if (is.character(eval_symbol)) {
+        symbol <- str2lang(paste0(names(dots)[i], " = ", eval_symbol))
+      }
+    }
+
+    # finally, we can evaluate expression and get values for new variables
+    new_variable <- try(with(data, eval(symbol)), silent = TRUE)
+
+    # successful, or any errors, like misspelled variable name?
+    if (inherits(new_variable, "try-error")) {
+      # in which step did error happen?
+      step_number <- switch(as.character(i),
+        "1" = "the first expression",
+        "2" = "the second expression",
+        "3" = "the third expression",
+        paste("expression", i)
+      )
+      step_msg <- paste0("There was an error in ", step_number, ".")
+      # try to find out which variable was the cause for the error
+      error_msg <- attributes(new_variable)$condition$message
+      if (grepl("object '(.*)' not found", error_msg)) {
+        error_var <- gsub("object '(.*)' not found", "\\1", error_msg)
+        insight::format_error(
+          paste0(step_msg, " Variable \"", error_var, "\" was not found in the dataset."),
+          .misspelled_string(colnames(data), error_var, "Possibly misspelled?")
+        )
+      } else {
+        insight::format_error(paste0(step_msg, " ", insight::format_capitalize(error_msg), ". Possibly misspelled?"))
+      }
+    }
+
+    # give informative error when new variable doesn't match number of rows
+    if (!is.null(new_variable) && length(new_variable) != nrow(data) && (nrow(data) %% length(new_variable)) != 0) {
+      insight::format_error("New variable has not the same length as the other variables in the data frame.")
+    }
+
+    data[[names(dots)[i]]] <- new_variable
+  }
+
+  data
+}
+
+#' @export
+data_modify.grouped_df <- function(data, ...) {
+  # we need to evaluate dots here, and pass them with "do.call" to
+  # the data.frame method later...
+  dots <- match.call(expand.dots = FALSE)$`...`
+  # save attributes, convert to regular data frame
+  grps <- attr(data, "groups", exact = TRUE)[[".rows"]]
+  attr_data <- attributes(data)
+  data <- as.data.frame(data)
+
+  ## TODO This code is duplicated, taken from the .data.frame method
+  ## @etiennebacher can we move this into a separate function and ".dynEval()" still works?
```

Comment:
I'll take a look at this

Reply:
Note that for grouped data, variable attributes like labels are lost but that's not specific to `data_modify()`, for example `data_arrange()` has the same problem